sql: truncate trailing spaces for char column types

otan opened the pull request:
Not sure if we we should merge this considering the backwards incompatibility, @solongordon / @rohany curious for your thoughts.

----

cc @arulajmani as peer programmer!

Resolves #50015

Release note (sql change, backward-incompatible change): The `CHAR`
column type will truncate the trailing space characters in line with Postgres.
This was previously not done. Existing stored `CHAR` entries with spaces at the
end of the char column type will no longer return rows with trailing space
characters - use the `LIKE` query to find and modify these rows.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50492)
<!-- Reviewable:end -->

rohany commented on pkg/sql/logictest/testdata/logic_test/collatedstring:
```diff
@@ -368,3 +368,18 @@ CREATE TABLE t46570(c0 BOOL, c1 STRING COLLATE en);
+
+subtest regression_50015
+
+statement ok
```
can you add a comment about what this test is doing + what does the output mean?

rohany commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -819,7 +819,7 @@ var builtins = map[string]builtinDefinition{
 	"repeat": makeBuiltin(defProps(),
 		tree.Overload{
-			Types:      tree.ArgTypes{{"input", types.String}, {"repeat_counter", types.Int}},
+			Types:      tree.ArgTypes{{"input", types.String}, {"repeat_counter", types.Int}},
```
What changed here?

otan replied:
mistake

rohany commented on pkg/sql/sem/tree/testdata/eval/cast:
```diff
@@ -1111,3 +1111,44 @@ eval
 ARRAY['he',e'a\U0001F41B']
+
+# Chars should truncate trailing space characters.
+# eval
```
eval is uncommented

rohany commented on pkg/sql/sqlbase/column_type_properties.go:
```diff
@@ -51,11 +52,23 @@ func AdjustValueToColumnType(
 			sv = v.Contents
 		}
 
+		if typ.Oid() == oid.T_bpchar {
```
This is everywhere -- is there not a function we can pull this and other common string logic out into?

otan replied:
everywhere is slightly different, unfortunately :(

rohany commented on pkg/sql/logictest/testdata/logic_test/collatedstring:
```diff
@@ -368,3 +368,32 @@ CREATE TABLE t46570(c0 BOOL, c1 STRING COLLATE en);
+subtest regression_50015
+
+query T
+SELECT t FROM ( VALUES
```
can you format this with sqlfum.pt

rohany commented on pkg/sql/logictest/testdata/logic_test/collatedstring:
```diff
@@ -368,3 +368,32 @@ CREATE TABLE t46570(c0 BOOL, c1 STRING COLLATE en);
+hello t
+
+statement ok
+CREATE TABLE t50015(id int PRIMARY KEY, a char(100), b char(100) COLLATE en); INSERT INTO t50015 VALUES
```
[nit]: new line

rohany commented on pkg/sql/logictest/testdata/logic_test/collatedstring:
```diff
@@ -368,3 +368,32 @@ CREATE TABLE t46570(c0 BOOL, c1 STRING COLLATE en);
+query ITITI
+SELECT id, a, length(a), b, length(b::string) FROM t50015 ORDER BY id ASC
+----
+1  hello                                                                                                 5   hello        5
```
Sorry, I'm still not sure what I'm looking for here. So `a` is padded all the way out to 100, even though its length is 5?

otan replied:
nah, just trailing spaces gets deleted.
5 is good, because 'hello 6' is 6.

otan commented:
does the backwards compatibility described sound ok to you too, @rohany?

rohany commented:
yeah

otan committed:
sql: truncate trailing spaces for char column types

cc @arulajmani as peer programmer!

Release note (sql change, backward-incompatible change): The `CHAR`
column type will truncate the trailing space characters in line with Postgres.
This was previously not done. Existing stored `CHAR` entries with spaces at the
end of the char column type will no longer return rows with trailing space
characters - use the `LIKE` query to find and modify these rows.

otan commented:
ok! tftr!

bors r=rohany
